private/model/api/example.go

```diff
@@ -0,0 +1,342 @@
+}
+
+var exampleFuncMap = template.FuncMap{
+	"commentify":       commentify,
+	"wrap":             wrap,
+	"generateFunction": generateExample,
+	"generateTypes":    generateTypes,
+}
+
+var exampleCustomizations = map[string]template.FuncMap{}
+
+var exampleTmpls = template.Must(template.New("example").Funcs(exampleFuncMap).Parse(`
+{{ generateTypes . }}
+{{ commentify (wrap .Title 80 false) }}
+//
+{{ commentify (wrap .Description 80 false) }}
+func Example{{ .API.StructName }}_{{ .OperationName }}_{{.Index}}() {
+	svc := {{ .API.PackageName }}.New(session.New())
+
+	{{ generateFunction . }}
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+				{{ range $_, $ref := .Operation.ErrorRefs -}}
+				{{ if not ($.HasVisitedError $ref) -}}
+			case {{ .API.PackageName }}.{{ $ref.Shape.ErrorCodeName }}:
+				fmt.Println({{ .API.PackageName }}.{{ $ref.Shape.ErrorCodeName }}, aerr.Error())
+				{{ end -}}
+				{{ end -}}
+			default:
+				fmt.Println(aerr.Error())
+			}
+		} else {
+			// Print the error, cast err to awserr.Error to get the Code and
+			// Message from an error.
+			fmt.Println(err.Error())
+		}
+		return
+	}
+
+	fmt.Println(result)
+}
+`))
+
+// Names will return the name of the example. This will also be the name of the operation
+// that is to be tested.
+func (exs Examples) Names() []string {
+	names := make([]string, 0, len(exs))
+	for k := range exs {
+		names = append(names, k)
+	}
+
+	sort.Strings(names)
+	return names
+}
+
+func (exs Examples) GoCode() string {
+	buf := bytes.NewBuffer(nil)
+	for _, examples := range exs {
+		for i, ex := range examples {
+			ex.Index = fmt.Sprintf("%c", rune(i+int('a')))
+			buf.WriteString(util.GoFmt(ex.GoCode()))
+			buf.WriteString("\n")
+		}
+	}
+	return buf.String()
+}
+
+// ExampleCode will generate the example code for the given Example shape.
+func (ex Example) GoCode() string {
+	var buf bytes.Buffer
+	m := exampleFuncMap
+	if fMap, ok := exampleCustomizations[ex.API.PackageName()]; ok {
+		m = fMap
+	}
+	tmpl := exampleTmpls.Funcs(m)
+	if err := tmpl.ExecuteTemplate(&buf, "example", &ex); err != nil {
+		panic(err)
+	}
+
+	return strings.TrimSpace(buf.String())
+}
+
+func generateExample(ex Example) string {
+	buf := bytes.NewBuffer(nil)
+	buf.WriteString(fmt.Sprintf("input := &%s.%s {\n", ex.API.PackageName(), ex.Operation.InputRef.Shape.ShapeName))
+	if ex.Operation.HasInput() {
+		buf.WriteString(buildShape(&ex.Operation.InputRef, ex.Input, false))
+	}
+	buf.WriteString("}\n")
+	buf.WriteString(fmt.Sprintf("\nresult, err := svc.%s(input)\n", ex.OperationName))
+	return buf.String()
+}
+
+// generateTypes will generate no types for default examples, but customizations may
+// require their own defined types.
+func generateTypes(ex Example) string {
+	return ""
+}
+
+// buildShape will recursively build the referenced shape based on the json object
+// provided.
+func buildShape(ref *ShapeRef, shapes map[string]interface{}, isMap bool) string {
```

Comment:
I think this function can probably be split into a few components. specifically the components based on type of the shape value.